---
title: "BEAM-Ab Analysis"
format: html
editor: visual
---

## Loading Libraries

```{r}
library(Seurat)
library(scRepertoire)
library(patchwork)
library(ggplot2)
library(viridis)
library(dplyr)
```

## Loading Single-Cell Data

```{r}
pbmc.data <- Read10X(data.dir = "./data/BEAM.AB/10k/raw_feature_bc_matrix")

pbmc <- CreateSeuratObject(counts = pbmc.data[[1]], 
                           project = "pbmc10k", 
                           min.cells = 3, 
                           min.features = 200)

#Attaching the BEAM-Ab data to single-cell object
BEAM <- pbmc.data[[2]][,colnames(pbmc.data[[2]]) %in% rownames(pbmc@meta.data)]
BEAM <- CreateAssayObject(counts = BEAM)
pbmc[["BEAM"]] <- BEAM
```

## Processing Single-Cell Object

```{r}
pbmc[["percent.mt"]] <- PercentageFeatureSet(pbmc, pattern = "^mt-")
VlnPlot(pbmc, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)

plot1 <- FeatureScatter(pbmc, feature1 = "nCount_RNA", feature2 = "percent.mt")
plot2 <- FeatureScatter(pbmc, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
plot1 + plot2

all.genes <- rownames(pbmc)

pbmc <- pbmc %>% 
          subset(subset = nFeature_RNA > 200 & nFeature_RNA < 3500 & percent.mt < 10) %>%
          NormalizeData(verbose = FALSE) %>%
          FindVariableFeatures(verbose = FALSE) 

#Removing IG genes
VariableFeatures(pbmc) <- VariableFeatures(pbmc)[-grep("^Ig[hlk][vdgcagm]", VariableFeatures(pbmc))]


pbmc <- pbmc %>%
          ScaleData(vars.to.regress = "percent.mt", 
                    features = all.genes, 
                    verbose = FALSE) %>%
          RunPCA(features = VariableFeatures(pbmc), 
                 verbose = FALSE)


ElbowPlot(pbmc)

```

Clustering and Dimensional Reduction

```{r}
pbmc <- pbmc %>%
  FindNeighbors(dims = 1:10) %>%
  FindClusters(algorithm = 4) %>%
  RunUMAP(dims = 1:10)

DimPlot(pbmc, label = TRUE) + NoLegend()
```

```{r}
pbmc.markers <- FindAllMarkers(pbmc, only.pos = TRUE)
pbmc.markers %>%
    group_by(cluster) %>%
    dplyr::filter(avg_log2FC > 1) %>%
    slice_max(order_by = avg_log2FC, n = 5)
```

## Loading BCR Sequences

```{r}
contigs <- read.csv("./data/BEAM.AB/10k/10k_BEAM-Ab_Mouse_B12_HEL_spikein_5pv2_Multiplex_vdj_b_all_contig_annotations.csv")
combined.BCR <- combineBCR(contigs,
                           samples = "S1")

combined.BCR <- lapply(combined.BCR, function(x) {
              x$barcode <- stringr::str_split(x$barcode, "_", simplify = TRUE)[,2]
              x
})

pbmc <- combineExpression(combined.BCR, pbmc)

DimPlot(pbmc, group.by = "cloneSize") + 
  scale_color_viridis(discrete = TRUE, direction = -1)
```

```{r}

plot1 <- FeatureScatter(pbmc, feature1 = "nCount_RNA", feature2 = "nCount_BEAM")
plot2 <- FeatureScatter(pbmc, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
plot1 + plot2

DefaultAssay(pbmc) <- "BEAM"

pbmc <- pbmc %>%
           ScaleData()

FeatureScatter(pbmc, "Anti-Hen-Egg-Lysozyme", "gp120")
FeatureScatter(pbmc, "Anti-Hen-Egg-Lysozyme", "negative-control")
FeatureScatter(pbmc, "gp120", "negative-control")
```

Method for Calling Ag+ Cells

Running Mouse Derived Ibex Models

Reduction comparing RNA, Ibex, RNA + Ibex as a function of Ag specificity
