---
title: "BEAM-Ab Analysis"
format: html
editor: visual
---

## Loading Libraries

```{r}
library(Seurat)
library(scRepertoire)
library(patchwork)
library(ggplot2)
library(viridis)
library(dplyr)
library(SingleR)
library(Azimuth)
library(scDblFinder)
library(harmony)
library(RColorBrewer)
library(Nebulosa)

HPCA <- HumanPrimaryCellAtlasData()
```

## Loading and Filtering Single-Cell Data

```{r}
SC.Files <- list.files("./data/BEAM.AB/Dugan2021/Recovered/")
lapply(SC.Files, function(x) {
      tmp <- Read10X(data.dir = paste0("./data/BEAM.AB/Dugan2021/Recovered/", x))
      
      
      SeuratObj <-  CreateSeuratObject(counts = tmp[[1]], 
                               project = x, 
                               min.cells = 3, 
                               min.features = 200)
      
      SeuratObj$nCount_RNA <- colSums(SeuratObj@assays$RNA$counts)
      SeuratObj$nFeature_RNA <- colSums(SeuratObj@assays$RNA$counts != 0)
      
      #Attaching the BEAM-Ab data to single-cell object
      BEAM <- tmp[[2]][,colnames(tmp[[2]]) %in% rownames(SeuratObj@meta.data)]
      BEAM <- CreateAssayObject(counts = BEAM)
      SeuratObj[["BEAM"]] <- BEAM
      SeuratObj$nCount_BEAM <- colSums(SeuratObj@assays$BEAM$counts)
      SeuratObj$nFeature_BEAM <- colSums(SeuratObj@assays$BEAM$counts != 0)
      
      #Adding Sample Acession to barcodes
      new.names <- paste0(x, "_", colnames(SeuratObj))
      SeuratObj <- RenameCells(SeuratObj, new.names = new.names)
      
      #Calculating percentage of Genes
      SeuratObj[["percent.ribo"]] <- PercentageFeatureSet(SeuratObj, pattern = "^RPS|RPL-")
      SeuratObj[["percent.mt"]] <- PercentageFeatureSet(SeuratObj, pattern = "^MT-")
      
      #Filtering the Data Set for QC Metrics
      standev <- sd(log(SeuratObj$nFeature_RNA))*2.5 #cutting off above standard deviation of 2.5
      mean <- mean(log(SeuratObj$nFeature_RNA))
      cut <- round(exp(standev+mean))
      SeuratObj <- subset(SeuratObj, subset = percent.mt < 10 & nFeature_RNA < cut)
    
      ###########################################
      #Estimate Doublets for Each Sequencing Run
      ############################################
      sce <- as.SingleCellExperiment(SeuratObj)
      sce <- scDblFinder(sce)
      doublets <- data.frame(db.class = sce$scDblFinder.class, db.score = sce$scDblFinder.score)
      rownames(doublets) <- rownames(sce@colData)
      SeuratObj <- AddMetaData(SeuratObj, doublets)
  
      #######################
      #Adding BCR information
      #######################
      
      contigs <- read.csv(paste0("./data/BEAM.AB/Dugan2021/Recovered/", x, "/filtered_contig_annotations.csv.gz"))
      combined.BCR <- combineBCR(contigs,
                               samples = x)
      SeuratObj <- combineExpression(combined.BCR, SeuratObj)
    
      ###########################################
      #Seurat Azimuth Annotation
      ############################################
      AziObject <- NormalizeData(SeuratObj , verbose = FALSE)
      AziObject <- ScaleData(AziObject , verbose = FALSE)
      AziObject <- FindVariableFeatures(AziObject , nfeatures = 2500)
      VariableFeatures(AziObject ) <- VariableFeatures(AziObject )[-grep("^IG[HLK][VDJCGAMD]", VariableFeatures(AziObject ))]
      AziObject <- RunPCA(AziObject , verbose = FALSE)
      AziObject <- RunUMAP(AziObject , dims = 1:30, verbose = FALSE)
      AziObject <- FindNeighbors(object = AziObject , 
                           features = VariableFeatures(AziObject ), 
                           verbose = FALSE)
     
      AziObject <- RunAzimuth(AziObject, 
                              reference = "pbmcref",
                              verbose = FALSE)
     
      SeuratObj <- AddMetaData(SeuratObj, AziObject[[]])
        
      #############################################
      #Singler Annotation of Cell Types
      #############################################
        
      com.res1 <- SingleR(sce, ref=HPCA, labels=HPCA$label.fine, assay.type.test=1)
    
      df <- data.frame("labels" = com.res1$labels, "pruned.labels" = com.res1$pruned.labels)
      rownames(df) <- rownames(com.res1)
      colnames(df) <- paste0("HPCA.", colnames(df))
      SeuratObj <- AddMetaData(SeuratObj,  df)
      
      
      ####################################
      #Final Filter: Only B cells with BCR
      ####################################
      Bcells <- rownames(SeuratObj[[]])[which(SeuratObj[[]]$predicted.celltype.l1 == "B" & !is.na(SeuratObj$CTaa))]
      SeuratObj <- subset(SeuratObj, cells = Bcells)
      
      #Filtering out any cell without both IGH/L chain
      Bcells <- rownames(SeuratObj[[]])[!grepl("NA_|_NA", SeuratObj$CTaa)]
      SeuratObj <- subset(SeuratObj, cells = Bcells)
      
      SeuratObj
}) -> SeuratList

SeuratMerge <- merge(SeuratList[[1]], SeuratList[-1])
rm(SeuratList)
gc()
saveRDS(SeuratMerge, "./data/processed/Dugan2021_BEAM.ab_Recovered_Seurat.rds")
```

## Merging and Dimensional Reduction

```{r}

SeuratMerge <- JoinLayers(SeuratMerge)
DefaultAssay(SeuratMerge) <- "RNA"

all.genes <- rownames(SeuratMerge@assays$RNA$counts)
SeuratMerge  <- NormalizeData(SeuratMerge)
SeuratMerge  <- FindVariableFeatures(SeuratMerge, nfeatures = 2500)
VariableFeatures(SeuratMerge) <- VariableFeatures(SeuratMerge)[-grep("^IG[HLK][VDJCAGM]", VariableFeatures(SeuratMerge))]

SeuratMerge  <- SeuratMerge %>%
                  ScaleData(vars.to.regress = "percent.mt", 
                            features = all.genes, 
                            verbose = FALSE) %>%
                  RunPCA(features = VariableFeatures(SeuratMerge), 
                         verbose = FALSE)
SeuratMerge <- RunHarmony(SeuratMerge, "orig.ident")

SeuratMerge <- SeuratMerge %>%
                  FindNeighbors(reduction = "harmony") %>%
                  FindClusters(algorithm = 4) 
        
SeuratMerge <- RunUMAP(SeuratMerge, 
                       reduction = "harmony", 
                       dims = 1:10)



saveRDS(SeuratMerge, "./data/processed/Dugan2021_BEAM.ab_Recovered_Seurat.rds")
```

## Figure 3

```{r}
#################
#Visualizing UMAP
#################

SeuratMerge <- readRDS("./data/processed/Dugan2021_BEAM.ab_Recovered_Seurat.rds")

Cluster.Pal <- colorRampPalette(brewer.pal(11, "Paired"))(length(unique(SeuratMerge$seurat_clusters)))


plot3.A <- DimPlot(SeuratMerge, label = TRUE, repel = TRUE, label.size = 6) +
  scale_color_manual(values = Cluster.Pal) + 
  theme_void() + 
  NoLegend() 

ggsave("./outputs/Figure3/Figure3A.png", plot3.A, height = 4.5, width = 5, dpi = 300)

########################
#Visualizing BEAM Signal
########################

DefaultAssay(SeuratMerge) <- "BEAM"

SeuratMerge  <- NormalizeData(SeuratMerge, 
                              nomalization.method = "CLR",
                              margin = 1)
SeuratMerge  <- ScaleData(SeuratMerge,
                          use.umi = TRUE)


density.plots <- plot_density(SeuratMerge, 
                              features = rownames(SeuratMerge@assays$BEAM$counts)[c(5,3,4,2,1)], 
                              combine = FALSE, 
                              method = "wkde") 

for(i in seq_along(density.plots)) {
  density.plots[[i]] <- density.plots[[i]] + 
                                  labs(title = NULL) + 
                                  guides(color = F) + 
                                  theme_void()
}


plot3.B <- density.plots[[1]] + density.plots[[2]] + density.plots[[3]] + density.plots[[4]] + density.plots[[5]] + plot_layout(ncol = 5)
ggsave("./outputs/Figure3/Figure3B.png", plot3.B, height = 4.5, width = 25, dpi = 300)

###############################
#Assigning Epitope Specificity
##############################

#From [Dugan et Al 2021 Mehods: https://pubmed.ncbi.nlm.nih.gov/34022127/
#normalized antigen-probe binding signals, we arbitrarily set a threshold equal to 1 for all normalized probe signals to distinguish probe binding cells as “positive” or “negative.” 

#Only cells whose top hit probe value was at least two-fold greater than their second hit probe value were clustered into the top hit probe-specific group; others were clustered into the “multi-reactive” group that indicates non-specific cells. 

#For samples in which we included separate SARS2 spike and RBD oligo tags, we placed cells positive to both SARS2 spike and SARS2 RBD into the “spike” group.

epitopeMatrix <- as.matrix(SeuratMerge@assays$BEAM$scale.data[1:5,])

lapply(seq_len(ncol(epitopeMatrix)), function(x) {
  
  if(sort(epitopeMatrix[,x])[5] > 2*sort(epitopeMatrix[,x])[4] & sort(epitopeMatrix[,x])[5] > 0.8) {
    epi <- names(sort(epitopeMatrix[,x]))[5]
  } else {
    epi <- NA
  }
  epi
}) -> epitope.assignments
names(epitope.assignments) <- colnames(epitopeMatrix)
epitope.assignments <- as.data.frame(do.call(rbind, epitope.assignments))
colnames(epitope.assignments) <- "Epitope"
epitope.assignments$Epitope <- stringr::str_split(epitope.assignments$Epitope, "-", simplify = TRUE)[,2]
epitope.assignments$Epitope[epitope.assignments$Epitope == ""] <- NA
epitope.assignments$Epitope <- factor(epitope.assignments$Epitope, levels = c("NP", "OFR7", "ORF8", "RBD", "Spike"))

SeuratMerge <- AddMetaData(SeuratMerge, epitope.assignments)



plot3.C <- DimPlot(SeuratMerge, split.by = "Epitope", pt.size = 1.5) +
  scale_color_manual(values = Cluster.Pal) + 
  ggtitle(NULL) + 
  guides(color = F) + 
 theme(strip.background = element_blank(),
   strip.text.x = element_blank()) + 
  theme_void()

ggsave("./outputs/Figure3/Figure3C.png", plot3.C, height = 4.6, width = 25, dpi = 300)

```

## Processing Benisse

Here we need to preprocess the 10x outputs to match the [Benisse workflow](https://github.com/wooyongc/Benisse). Need to create 2 csv files: 1) an expression matrix and 2) a contig file with the contig ID and cdr3 AA sequence for IGH loci.

```{r}
library(Seurat)
library(dplyr)

#####################
#Processing RNA Data
#####################

SC.Files <- list.files("./data/BEAM.AB/Dugan2021/Recovered/")
for(i in seq_along(SC.Files)) {
    tmp <- Read10X(data.dir = paste0("./data/BEAM.AB/Dugan2021/Recovered/", SC.Files[i]))
    tmp <- as.data.frame(tmp[[1]])
    write.csv(tmp, paste0("./data/Benisse/inputs/", SC.Files[i], "_exp.csv"))
}


#######################
#Processing Contig Data
#######################

for(i in seq_along(SC.Files)) {
  tmp <- read.csv(paste0("./data/BEAM.AB/Dugan2021/Recovered/", SC.Files[i], "/filtered_contig_annotations.csv.gz"))
  tmp <- tmp[tmp$productive == "True" & tmp$chain == "IGH",]
  tmp <- tmp %>%
          group_by(barcode) %>%
          slice_max(order_by = reads, n=1) %>%
          as.data.frame()
  
  tmp <- tmp %>%
          group_by(barcode) %>%
          slice_max(order_by = umis, n=1) %>%
          as.data.frame()
          
  tmp <- data.frame(row.names = tmp$barcode,
                    contigs = tmp$contig_id, 
                    cdr3 = tmp$cdr3)
  
  rownames(tmp) <- stringr::str_replace_all(rownames(tmp), "-", ".")
  
  write.csv(tmp, paste0("./data/Benisse/inputs/", SC.Files[i], ".csv"))
}


```

### Step 1: BCR Encoding

```{python eval = FALSE}
# Define the range of samples
start_sample = 5231092
end_sample = 5231101

# Define the base command and file paths
base_command = "python3 AchillesEncoder.py"
input_base_path = "/Users/nick/Documents/GitHub/Ibex.manuscript/data/Benisse/inputs/"
output_base_path = "/Users/nick/Documents/GitHub/Ibex.manuscript/data/Benisse/encoded/"
cuda_flag = "--cuda False"

# Loop through the sample numbers and generate the command line scripts
for sample_number in range(start_sample, end_sample + 1):
    sample_label = f"GSM{sample_number}"
    input_file = f"{input_base_path}{sample_label}.csv"
    output_file = f"{output_base_path}{sample_label}_encoded.csv"
    command = f"{base_command} --input_data {input_file} --output_data {output_file} {cuda_flag}"
    print(command)
```
