---
title: "Ibex Manuscript Analysis"
author: "Nick Borcherding"
date: "8/18/2022"
output: html_document
---

```{r}
suppressPackageStartupMessages(library(Seurat))
suppressPackageStartupMessages(library(ggplot2))
suppressPackageStartupMessages(library(dplyr))
suppressPackageStartupMessages(library(gridExtra))
suppressPackageStartupMessages(library(viridis))
suppressPackageStartupMessages(library(RColorBrewer))
suppressPackageStartupMessages(library(Matrix))
suppressPackageStartupMessages(library(stringr))
suppressPackageStartupMessages(library(Ibex))
suppressPackageStartupMessages(library(mumosa))
```

```{r}
library(reticulate)
Sys.setenv(RETICULATE_PYTHON = "/Users/ncborch/Library/r-miniconda-arm64/envs/r-reticulate/bin/python")
use_condaenv(condaenv = "r-reticulate", required = TRUE)
```

# Ibex performance on naive repertoire

## Loading BCR sequences 

BCR sequences being loaded are sequences never seen by the autoencoder. The R folder contains stripped down version of the Ibex encoders to allow for the comparisons without the need for single-cell data.

```{r, eval=FALSE}
source("./R/utils.R")
source("./R/auto.encoder.R")

#CDR3 sequences that the autoencoders have never seen before
unique.heavy <- readRDS("./data/naive.heavy.rds")
unique.light <- readRDS("./data/naive.heavy.rds")


######################
#Setting up the loops
#####################
#Testing both chains, 3 AA property encoders, testing time for chains, and bootstrapping everything together
set.seed(42)
chain.list <- list(unique.heavy, unique.light)
names(chain.list) <- c("Heavy", "Light")
num.chains <- c(10,100,1000,10000, 100000)
boot.straps <- 1:10
AA.list <- c("KF", "AF", "OHE")

final.output <- NULL
chain.output <- NULL
int.output <- NULL
boot.output <- NULL
for(i in seq_along(chain.list)) {
    for(j in seq_along(num.chains)) {
      tmp.chain <- chain.list[[i]][sample(length(chain.list[[i]]), num.chains[j])]
      for(k in seq_along(boot.straps)) {
        for(l in seq_along(AA.list)) {
          start_time <- Sys.time()
          out.tmp <- auto.encoder(tmp.chain, 
                         chain = names(chain.list)[i], 
                         AA.properties = AA.list[l])
          end_time <- Sys.time()
          runtime <- end_time - start_time
          out <- c(nrow(unique(out.tmp)), runtime, names(chain.list)[i], AA.list[l], num.chains[j], boot.straps[k])
          int.output <- rbind(int.output, out)
        }
        boot.output <- rbind(boot.output, int.output)
        int.output <- NULL
      }
      chain.output <- rbind(chain.output, boot.output)
      boot.output <- NULL
    }
  final.output <- rbind(final.output, chain.output)
}
final.output <- as.data.frame(final.output)
colnames(final.output) <- c("unique.cdr3", "runtime", "chain", "property", "num.cdr3.tested", "boot.strap")

saveRDS(final.output, "./output/Benchmarking.results.rds")
```


## Graphing the Benchmarking results

```{r}
final.output <- readRDS("./output/Benchmarking.results.rds")

plot1 <- ggplot(final.output, aes(x = as.numeric(num.cdr3.tested), y = as.numeric(unique.cdr3), group = interaction(as.numeric(num.cdr3.tested),property))) + 
  geom_smooth(method = "lm", color = "black", size = 0.5) + 
  geom_jitter(aes(color = property)) + 
  scale_x_log10() + 
  scale_y_log10() + 
  xlab("cdr3 sequences tested") + 
  ylab("unique encoded values") + 
  scale_color_viridis(discrete = TRUE)+
  facet_grid(.~chain) + 
  theme_bw()

plot2 <- ggplot(final.output, aes(x = as.numeric(num.cdr3.tested), y = as.numeric(runtime), group = property)) + 
  geom_smooth(method = "loess", size = 0.5, aes(color = property)) + 
  geom_point(aes(color = property)) + 
   xlab("cdr3 sequences tested") + 
  ylab("time (s)") + 
  scale_x_log10() + 
  scale_color_viridis(discrete = TRUE)+
  facet_grid(.~chain) + 
  theme_bw()

plot1 + plot2 + plot_layout(ncol = 1)
ggsave("./output/benchmarking.pdf", height = 7, width = 7)
```


# The Data Set

To show the multiple options of Ibex, the example data is derived from [this manuscript (https://pubmed.ncbi.nlm.nih.gov/33891889/) - multimodal single-cell characterization of COVID19-associated multisystem inflammatory syndrome in children. The data example built into the package (ibex_example) is derived from randomly sampling cells from Patient 1.

## Preprocessing Data

```{r, eval=FALSE}
suppressMessages(library(scRepertoire))
##################################
#scRNA/ADT loading and processing
#################################
tmp <-  Read10X("./data/GSM5073055_P1.1_filtered_feature_bc_matrix")

MIS.sample <- CreateSeuratObject(counts = tmp$`Gene Expression`)
rownames(tmp$`Antibody Capture`) <- stringr::str_remove_all(rownames(tmp$`Antibody Capture`), "anti_human_")
rownames(tmp$`Antibody Capture`) <- stringr::str_remove_all(rownames(tmp$`Antibody Capture`), "anti_mousehuman_")
rownames(tmp$`Antibody Capture`) <- substr(rownames(tmp$`Antibody Capture`), 6, nchar(rownames(tmp$`Antibody Capture`)))

adt_assay <- CreateAssayObject(counts = tmp$`Antibody Capture`)


MIS.sample[["ADT"]] <- adt_assay
MIS.sample <- subset(MIS.sample, subset = nFeature_RNA > 100) 
MIS.sample  <- RenameCells(object = MIS.sample , new.names = paste0("MIS.sample_", rownames(MIS.sample[[]])))
MIS.sample[["mito.genes"]] <- PercentageFeatureSet(MIS.sample, pattern = "^MT-")
    
#Filtering step
standev <- sd(log(MIS.sample$nFeature_RNA))*2.5 #cutting off above standard deviation of 2.5
mean <- mean(log(MIS.sample$nFeature_RNA))
cut <- round(exp(standev+mean))
MIS.sample <- subset(MIS.sample, subset = mito.genes < 10 & nFeature_RNA < cut)

#Processing and Adding Contig Info
contigs <- read.csv("./data/GSM5073091_PBMC_P1.1_MIS-C_Severe_BCR_filtered_contig_annotations.csv.gz")
clones <- combineBCR(contigs, samples = "MIS.sample", removeNA = TRUE)
MIS.sample <- combineExpression(clones, MIS.sample, cloneCall="aa")

#Subset only B cells (by contigs)
MIS.sample$BCR.recoverd <- "No"
MIS.sample$BCR.recoverd[!is.na(MIS.sample$CTaa)] <- "Yes"
MIS.sample <- subset(MIS.sample, BCR.recoverd == "Yes")

#Processing RNA
DefaultAssay(MIS.sample) <- 'RNA'
MIS.sample <- NormalizeData(MIS.sample) %>% FindVariableFeatures() %>% 
  quietBCRgenes() %>% ScaleData() %>% RunPCA(verbose = FALSE)

#Processing ADT
DefaultAssay(MIS.sample) <- 'ADT'
VariableFeatures(MIS.sample) <- rownames(MIS.sample[["ADT"]])
MIS.sample <- NormalizeData(MIS.sample, normalization.method = 'CLR', margin = 2) %>% 
  ScaleData() %>% RunPCA(reduction.name = 'apca')

saveRDS("/data/Ibex_FullExample.rds")
```


```{r}
dir.create("./output")

SeuratObj<- readRDS("./data/Ibex_FullExample.rds")

############################
#Adding Vgenes to meta data
###########################

SeuratObj$Heavy.V <- str_split(SeuratObj$CTgene, "[.]", simplify = TRUE)[,1]
SeuratObj$Light.V <- str_split(str_split(SeuratObj$CTgene, "[_]", simplify = TRUE)[,2], "[.]", simplify = TRUE)[,1]

##################
#Running Ibex
################
SeuratObj <- runIbex(SeuratObj, 
                    chains = "Heavy",
                    AA.properties = "KF", 
                    reduction.name = "Ibex.H.KF")

SeuratObj <- runIbex(SeuratObj, 
                    chains = "Light",
                    AA.properties = "KF", 
                    reduction.name = "Ibex.L.KF")


heavy.plot.V <- DimPlot(SeuratObj, reduction = "Ibex.H.KF", group.by = "Heavy.V") + 
                  scale_color_viridis(discrete = TRUE) + 
                  theme_void() + 
                  guides(fill = "none") + 
                  theme(plot.title = element_blank()) +
                  NoLegend()

light.plot.V <- DimPlot(SeuratObj, reduction = "Ibex.L.KF", group.by = "Light.V") + 
                  scale_color_viridis(discrete = TRUE) + 
                  theme_void() + 
                  guides(fill = "none") + 
                  theme(plot.title = element_blank()) +
                  NoLegend()

heavy.plot.V 
ggsave("./output/Ibex_heavy_vgenes.pdf", height = 3.5, width = 3.5)
light.plot.V 
ggsave("./output/Ibex_light_vgenes.pdf", height = 3.5, width = 3.5)
```

## Reducing the data set to clonotypes

```{r}
conga <- CoNGAfy(SeuratObj, 
                 assay = c("RNA", "ADT"))

#Get count of number of clones
meta <- SeuratObj[[]]
clone.count <- meta %>%
  group_by(CTaa) %>%
  summarise(n = n()) %>%
  as.data.frame()

ID <- clone.count$CTaa
clone.count <- data.frame(n = clone.count[,2])
rownames(clone.count) <- ID
conga <- AddMetaData(conga, clone.count)

DefaultAssay(conga) <- 'RNA'
conga <- NormalizeData(conga) %>% 
  FindVariableFeatures() %>% 
  quietBCRgenes() %>% 
  ScaleData() %>% 
  RunPCA(verbose = FALSE)

#Processing ADT
DefaultAssay(conga) <- 'ADT'
IG.ADT <- grep("Ig",  rownames(conga[["ADT"]]))
VariableFeatures(conga) <- rownames(conga[["ADT"]])[-IG.ADT] #Removing Immunoglobulins from DimRed

conga <- NormalizeData(conga, normalization.method = 'CLR', margin = 2) %>% 
  ScaleData() %>% 
  RunPCA(reduction.name = 'apca')

##################
#Running Ibex
################
conga <- runIbex(conga, 
                    chains = "Heavy",
                    AA.properties = "KF", 
                    reduction.name = "Ibex.H.KF")

conga <- runIbex(conga, 
                    chains = "Light",
                    AA.properties = "KF", 
                    reduction.name = "Ibex.L.KF")



###################
#Trimodal Embedding
###################
output <- rescaleByNeighbors(
                  list(conga[["Ibex.H.KF"]]@cell.embeddings, 
                      conga[["pca"]]@cell.embeddings, 
                      conga[["apca"]]@cell.embeddings), 
                  k = 50)

conga[["Trimodal.Embed"]] <- CreateDimReducObject(
                                embeddings = as.matrix(output),
                                stdev = rep(0, ncol(output)),
                                key = "Combined",
                                jackstraw = NULL,
                                misc = list())

conga <- FindNeighbors(conga, 
                       reduction = "Trimodal.Embed", 
                       dims = 1:130, 
                       graph.name = c("Ibex_nn", "Ibex_snn"))
conga <- FindClusters(conga,
                      graph.name = "Ibex_nn",
                      algorithm = 4)

#######################################
#UMAP Comparisons between each approach
#######################################

conga <- RunUMAP(conga, 
                     reduction = 'pca', 
                     dims = 1:30, 
                     assay = 'RNA', 
                     reduction.name = 'rna.umap', 
                     reduction.key = 'rnaUMAP_')

conga <- RunUMAP(conga, 
                     reduction = 'apca', 
                     dims = 1:30, 
                     assay = 'ADT', 
                     reduction.name = 'adt.umap', 
                     reduction.key = 'adtUMAP_')

conga <- RunUMAP(conga, 
                     reduction = 'Ibex.H.KF', 
                     dims = 1:30, 
                     reduction.name = 'heavy.umap', 
                     reduction.key = 'heavyUMAP_')

conga <- RunUMAP(conga, 
                     reduction = 'Ibex.L.KF', 
                     dims = 1:30, 
                     reduction.name = 'light.umap', 
                     reduction.key = 'lightUMAP_')

conga <- RunUMAP(conga, 
                     reduction = 'Trimodal.Embed', 
                     dims = 1:130, 
                     reduction.name = 'trimodal.umap', 
                     reduction.key = 'trimodalUMAP_')

plot.rna <- DimPlot(conga, reduction = "rna.umap") + NoLegend()
plot.adt <- DimPlot(conga, reduction = "adt.umap") + NoLegend()
plot.heavy <- DimPlot(conga, reduction = "heavy.umap") + NoLegend()
plot.combined <- DimPlot(conga, reduction = "trimodal.umap") + NoLegend()

###############
#Plotting
#############

plot.rna + 
  scale_color_tableau() + 
  theme_void() + 
  NoLegend()
ggsave("./output/clonotype_RNA_umap.pdf", height = 3.5, width = 3.5)

plot.adt + 
  scale_color_tableau() + 
  theme_void() + 
  NoLegend()
ggsave("./output/clonotype_ADT_umap.pdf", height = 3.5, width = 3.5)

plot.heavy + 
  scale_color_tableau() + 
  theme_void() + 
  NoLegend()
ggsave("./output/clonotype_IgHeavy_umap.pdf", height = 3.5, width = 3.5)

plot.combined + 
  scale_color_tableau() + 
  theme_void() + 
  NoLegend()
ggsave("./output/clonotype_combined_umap.pdf", height = 3.5, width = 3.5)
  

saveRDS(conga, file = "./data/Ibex_ClonotypeRep.rds")
```

