---
title: "Ibex Manuscript Analysis"
author: "Nick Borcherding"
date: "4/14/2023"
output: html_document
---

```{r}
suppressPackageStartupMessages(library(Seurat))
suppressPackageStartupMessages(library(ggplot2))
suppressPackageStartupMessages(library(dplyr))
suppressPackageStartupMessages(library(gridExtra))
suppressPackageStartupMessages(library(viridis))
suppressPackageStartupMessages(library(RColorBrewer))
suppressPackageStartupMessages(library(Matrix))
suppressPackageStartupMessages(library(stringr))
suppressPackageStartupMessages(library(Ibex))
suppressPackageStartupMessages(library(mumosa))
suppressPackageStartupMessages(library(patchwork))
```

```{r}
library(reticulate)
#Sys.setenv(RETICULATE_PYTHON = "/Users/ncborch/Library/r-miniconda-arm64/envs/r-reticulate/bin/python")
use_condaenv(condaenv = "r-reticulate", required = TRUE)
```

# Ibex performance on naive repertoire

## Loading BCR sequences 

BCR sequences being loaded are sequences never seen by the autoencoder. The R folder contains stripped down version of the Ibex encoders to allow for the comparisons without the need for single-cell data.

```{r, eval=FALSE}
source("./R/utils.R")
source("./R/auto.encoder.R")

#CDR3 sequences that the autoencoders have never seen before
unique.heavy <- readRDS("./data/naive.heavy.rds")
unique.light <- readRDS("./data/naive.light.rds")


######################
#Setting up the loops
#####################
#Testing both chains, 3 AA property encoders, testing time for chains, and bootstrapping everything together
set.seed(42)
chain.list <- list(unique.heavy, unique.light)
names(chain.list) <- c("Heavy", "Light")
num.chains <- c(10,100,1000,10000, 100000)
boot.straps <- 1:10
AA.list <- c("KF", "AF", "OHE")

final.output <- NULL
chain.output <- NULL
int.output <- NULL
boot.output <- NULL
for(i in seq_along(chain.list)) {
    for(j in seq_along(num.chains)) {
      tmp.chain <- chain.list[[i]][sample(length(chain.list[[i]]), num.chains[j])]
      for(k in seq_along(boot.straps)) {
        for(l in seq_along(AA.list)) {
          start_time <- Sys.time()
          out.tmp <- auto.encoder(tmp.chain, 
                         chain = names(chain.list)[i], 
                         AA.properties = AA.list[l])
          end_time <- Sys.time()
          runtime <- end_time - start_time
          out <- c(nrow(unique(out.tmp)), runtime, names(chain.list)[i], AA.list[l], num.chains[j], boot.straps[k])
          int.output <- rbind(int.output, out)
        }
        boot.output <- rbind(boot.output, int.output)
        int.output <- NULL
      }
      chain.output <- rbind(chain.output, boot.output)
      boot.output <- NULL
    }
  final.output <- rbind(final.output, chain.output)
}
final.output <- as.data.frame(final.output)
colnames(final.output) <- c("unique.cdr3", "runtime", "chain", "property", "num.cdr3.tested", "boot.strap")
rownames(final.output) <- NULL

saveRDS(final.output, "./output/Benchmarking.results.rds")
```


## Graphing the Benchmarking results

```{r}
final.output <- readRDS("./output/Benchmarking.results.rds")

plot1 <- ggplot(final.output, aes(x = as.numeric(num.cdr3.tested), y = as.numeric(unique.cdr3), group = interaction(as.numeric(num.cdr3.tested),property))) + 
  geom_smooth(method = "lm", color = "black", size = 0.5) + 
  geom_jitter(aes(color = property)) + 
  scale_x_log10() + 
  scale_y_log10() + 
  xlab("cdr3 sequences tested") + 
  ylab("unique encoded values") + 
  scale_color_viridis(discrete = TRUE)+
  facet_grid(.~chain) + 
  theme_bw()

plot2 <- ggplot(final.output, aes(x = as.numeric(num.cdr3.tested), y = as.numeric(runtime), group = property)) + 
  geom_smooth(method = "loess", size = 0.5, aes(color = property)) + 
  geom_point(aes(color = property)) + 
   xlab("cdr3 sequences tested") + 
  ylab("time (s)") + 
  scale_x_log10() + 
  scale_color_viridis(discrete = TRUE)+
  facet_grid(.~chain) + 
  theme_bw()

plot1 + plot2 + plot_layout(ncol = 1)
ggsave("./output/benchmarking.pdf", height = 7, width = 7)
```


# The Data Set

To show the multiple options of Ibex, the example data is derived from [this manuscript (https://pubmed.ncbi.nlm.nih.gov/33891889/) - multimodal single-cell characterization of COVID19-associated multisystem inflammatory syndrome in children. The data example built into the package (ibex_example) is derived from randomly sampling cells from Patient 1.

## Preprocessing Data

```{r, eval=FALSE}
dir.create("./qc")
GEX.matrices <- list.dirs("./data/sequencingRuns", 
                           recursive = TRUE,
                           full.names = TRUE)
GEX.matrices <- GEX.matrices[grep("filtered_feature_bc_matrix", GEX.matrices)]
sample.names <- list.dirs("./data/sequencingRuns", recursive = FALSE)
sample.names <- stringr::str_split(sample.names, "[/]", simplify = TRUE)[,4]

seurat.list <- NULL
for(i in seq_along(sample.names)) {
  tmp <- Read10X(GEX.matrices[[i]])
  if(length(tmp) == 2) {
     s.object <- CreateSeuratObject(counts = tmp$`Gene Expression`, sample.names[i])
     #Removing extra prefixes on antibodies
     rownames(tmp$`Antibody Capture`) <- stringr::str_remove_all(rownames(tmp$`Antibody Capture`), "anti_human_")
     rownames(tmp$`Antibody Capture`) <- stringr::str_remove_all(rownames(tmp$`Antibody Capture`), "anti_mousehuman_")
     rownames(tmp$`Antibody Capture`) <- substr(rownames(tmp$`Antibody Capture`), 6, nchar(rownames(tmp$`Antibody Capture`)))
  
     adt_assay <- CreateAssayObject(counts = tmp$`Antibody Capture`)
     s.object[["ADT"]] <- adt_assay
  } else {
    s.object <- CreateSeuratObject(counts = tmp, project = sample.names[i])
  }
  s.object <- RenameCells(object = s.object , new.names = paste0(sample.names[i], "_", rownames(s.object[[]])))
  s.object <- subset(s.object, subset = nFeature_RNA > 100) 
  s.object[["percent.mt"]] <- PercentageFeatureSet(s.object, pattern = "^MT-")
  s.object[["percent.rb"]] <- PercentageFeatureSet(s.object, pattern = "^RBS|RPL")
  
  p1 <- VlnPlot(object = s.object, features = c("nCount_RNA")) + theme(legend.position = "none")
  p2 <- VlnPlot(object = s.object, features = c("nFeature_RNA")) + theme(legend.position = "none")
  p3 <- VlnPlot(object = s.object, features = c("percent.mt")) + theme(legend.position = "none")
    
  
  p1 + p2 + p3+ plot_layout(ncol = 3)
  ggsave(paste0("./qc/", sample.names[i], ".pdf"), height = 8, width=12)
    
  ###########################
  #Here is the filtering step
  ############################
  standev <- sd(log(s.object$nCount_RNA))*2 #cutting off above standard deviation of 2
  mean <- mean(log(s.object$nCount_RNA))
  cut <- round(exp(standev+mean))
  s.object <- subset(s.object, subset = percent.mt < 15 & nFeature_RNA < cut)

  seurat.list[[i]] <- s.object
}

seurat.merge <- merge(seurat.list[[1]], seurat.list[-1])


###########################
#Adding Contig Information
############################
  
contig.file <- list.files("./data/sequencingRuns/", 
                            pattern = "contig", 
                            full.names = TRUE, 
                            recursive = TRUE)

contigs <- lapply(contig.file, read.csv)
clones <- combineBCR(contigs, samples = sample.names, removeNA = TRUE)
seurat.merge <- combineExpression(clones, seurat.merge, cloneCall = "aa")
seurat.merge$BCR.recoverd <- "No"
seurat.merge$BCR.recoverd[!is.na(seurat.merge$CTaa)] <- "Yes"
seurat.merge <- subset(seurat.merge, BCR.recoverd == "Yes")
saveRDS(seurat.merge, "./data/Ibex_SCdata.rds")

###########################
#Integration Step
############################
#Only fraction has ADT - we will just do overall RNA for now.
"%!in%" <- Negate("%in%")
seurat.merge <- readRDS("./data/Ibex_SCdata.rds")
seurat.list <- SplitObject(seurat.merge, split.by = "orig.ident")
seurat.list <- lapply(X = seurat.list, function(x) {
  trans <- SCTransform(x,
                        vars.to.regress = c('percent.mt', 'percent.rb', "nCount_RNA"),
                        min_cells = 3, 
                        verbose = FALSE)
  trans
})
features <- SelectIntegrationFeatures(object.list = seurat.list, nfeatures = 3000)
features <- Ibex::quietBCRgenes(features)
seurat.list <- PrepSCTIntegration(object.list = seurat.list, anchor.features = features)

immune.anchors <- FindIntegrationAnchors(object.list = seurat.list, normalization.method = "SCT", anchor.features = features)
seurat.merge <- IntegrateData(anchorset = immune.anchors, normalization.method = "SCT")
seurat.merge <- RunPCA(seurat.merge, verbose = FALSE)
seurat.merge <- RunUMAP(seurat.merge, reduction = "pca", dims = 1:30)
saveRDS(seurat.merge, "./data/Ibex_SCdata.rds")
```


```{r}
dir.create("./output")

seurat.merge <- readRDS("./data/Ibex_SCdata.rds")

############################
#Adding genes to meta data
###########################

seurat.merge$Heavy.V <- str_split(seurat.merge$CTgene, "[.]", simplify = TRUE)[,1]
seurat.merge$Light.V <- str_split(str_split(seurat.merge$CTgene, "[_]", simplify = TRUE)[,2], "[.]", simplify = TRUE)[,1]

seurat.merge$Heavy.D <- str_split(seurat.merge$CTgene, "[.]", simplify = TRUE)[,2]

seurat.merge$Heavy.J <- str_split(seurat.merge$CTgene, "[.]", simplify = TRUE)[,3]
seurat.merge$Light.V <- str_split(str_split(seurat.merge$CTgene, "[_]", simplify = TRUE)[,2], "[.]", simplify = TRUE)[,2]

seurat.merge$Heavy.C <- str_split(str_split(seurat.merge$CTgene, "[.]", simplify = TRUE)[,4], "[_]", simplify = TRUE)[,1]
seurat.merge$Light.C <- str_split(str_split(seurat.merge$CTgene, "[_]", simplify = TRUE)[,2], "[.]", simplify = TRUE)[,3]

##################
#Running Ibex
################
seurat.merge  <- runIbex(seurat.merge, 
                    chains = "Heavy",
                    AA.properties = "KF", 
                    reduction.name = "Ibex.H.KF")

seurat.merge  <- runIbex(seurat.merge, 
                    chains = "Light",
                    AA.properties = "KF", 
                    reduction.name = "Ibex.L.KF")

seurat.merge  <- runIbex(seurat.merge, 
                    chains = "Heavy",
                    AA.properties = "AF", 
                    reduction.name = "Ibex.H.AF")

seurat.merge  <- runIbex(seurat.merge, 
                    chains = "Light",
                    AA.properties = "AF", 
                    reduction.name = "Ibex.L.AF")

seurat.merge  <- runIbex(seurat.merge, 
                    chains = "Heavy",
                    AA.properties = "OHE", 
                    reduction.name = "Ibex.H.OHE")

seurat.merge  <- runIbex(seurat.merge, 
                    chains = "Light",
                    AA.properties = "OHE", 
                    reduction.name = "Ibex.L.OHE")


seurat.merge$status <- ifelse(seurat.merge$orig.ident %in% c("HD1", "HD2", "HD3", "HD4", "HD5"), "Healthy", "MIS")
saveRDS(seurat.merge, "./data/Ibex_SCdata.rds")

cells <- which(seurat.merge[[]]$orig.ident %in% c("P1.1", "P2.1", "HD1", "HD4"))

DimPlot(seurat.merge[,cells], reduction = "Ibex.H.KF", group.by = "Heavy.V") + 
                  scale_color_viridis(discrete = TRUE) + 
                  theme_void() + 
                  guides(fill = "none") + 
                  theme(plot.title = element_blank()) +
                  NoLegend()
ggsave("./output/Ibex_heavy_vgenes_KF.pdf", height = 3.5, width = 3.5)

DimPlot(seurat.merge, reduction = "Ibex.L.KF", group.by = "Light.V") + 
                  scale_color_viridis(discrete = TRUE) + 
                  theme_void() + 
                  guides(fill = "none") + 
                  theme(plot.title = element_blank()) +
                  NoLegend()
ggsave("./output/Ibex_light_vgenes_KF.pdf", height = 3.5, width = 3.5)

#############################
#Supplemental Visualizations
#############################

DimPlot(seurat.merge[,cells], reduction = "Ibex.H.AF", group.by = "Heavy.V") + 
                  scale_color_viridis(discrete = TRUE) + 
                  theme_void() + 
                  guides(fill = "none") + 
                  theme(plot.title = element_blank()) +
                  NoLegend()
ggsave("./output/Ibex_heavy_vgenes_AF.pdf", height = 3.5, width = 3.5)

DimPlot(seurat.merge, reduction = "Ibex.L.AF", group.by = "Light.V") + 
                  scale_color_viridis(discrete = TRUE) + 
                  theme_void() + 
                  guides(fill = "none") + 
                  theme(plot.title = element_blank()) +
                  NoLegend()
ggsave("./output/Ibex_light_vgenes_AF.pdf", height = 3.5, width = 3.5)

DimPlot(seurat.merge[,cells], reduction = "Ibex.H.OHE", group.by = "Heavy.V") + 
                  scale_color_viridis(discrete = TRUE) + 
                  theme_void() + 
                  guides(fill = "none") + 
                  theme(plot.title = element_blank()) +
                  NoLegend()
ggsave("./output/Ibex_heavy_vgenes_OHE.pdf", height = 3.5, width = 3.5)


DimPlot(seurat.merge, reduction = "Ibex.L.OHE", group.by = "Light.V") + 
                  scale_color_viridis(discrete = TRUE) + 
                  theme_void() + 
                  guides(fill = "none") + 
                  theme(plot.title = element_blank()) +
                  NoLegend()
ggsave("./output/Ibex_light_vgenes_OHE.pdf", height = 3.5, width = 3.5)

```

# Call Somatic Hypermutation

## Calculating Clonal Lineages and Hypermutation

Unfortunately only cdr3 sequences are available for these data sets - in order to call clonally related clonotypes, we will use edit distances and VDJ genes to identify clonal groups. This step was actually performed during ```combineBCR()``` where normalized edit difference of 0.85 are grouped together. 

```{r}
seurat.merge$Light.group <- stringr::str_split(seurat.merge$CTstrict, "_", simplify = TRUE)[,2]
seurat.merge$Light.group <- ifelse(seurat.merge$Light.group %in% names(which(table(seurat.merge[,cells]$Light.group) >= 10)), seurat.merge$Light.group, NA)

seurat.merge$Heavy.group <- stringr::str_split(seurat.merge$CTstrict, "_", simplify = TRUE)[,1]
seurat.merge$Heavy.group <- ifelse(seurat.merge$Heavy.group %in% names(which(table(seurat.merge[,cells]$Heavy.group) >= 10)), seurat.merge$Heavy.group, NA)
```

# Ploting Latent Dimensions

```{r}

heavy <- c("Ibex.H.KF", "Ibex.H.AF","Ibex.H.OHE")
light <- c("Ibex.L.KF", "Ibex.L.AF","Ibex.L.OHE")
cells <- which(seurat.merge[[]]$orig.ident %in% c("P1.1", "P2.1", "HD1", "HD4"))
    
for (j in seq_along(heavy)) {
        Dim.plot.H <- DimPlot(seurat.merge[,cells], reduction = heavy[j], group.by = "Heavy.V") + 
                          scale_color_viridis(discrete = TRUE) + 
                          theme_void() + 
                          guides(fill = "none") + 
                          theme(plot.title = element_blank()) +
                          NoLegend()
        
        Dim.plot.Ident <- DimPlot(seurat.merge[,cells], reduction = heavy[j], group.by = "orig.ident") + 
                              scale_color_viridis(discrete = TRUE) + 
                              theme_void() + 
                              guides(fill = "none") + 
                              theme(plot.title = element_blank())
        tmp <- subset(seurat.merge, cells = rownames(seurat.merge[[]])[cells])
        
        #Calculate MDS   
        d <- dist(tmp@reductions[[heavy[j]]]@cell.embeddings)
        mds <- cmdscale(d = d, k = 2)
        colnames(mds) <- paste0("MDS_", 1:2)
        tmp[["mds"]] <- CreateDimReducObject(embeddings = mds, key = "MDS_", assay = DefaultAssay(tmp))
         
        MDS.plot.H <- DimPlot(tmp, reduction = "mds", group.by = "Heavy.V") + 
                          scale_color_viridis(discrete = TRUE) + 
                          theme_void() + 
                          guides(fill = "none") + 
                          theme(plot.title = element_blank()) +
                          NoLegend() 
        
         MDS.plot.Ident <- DimPlot(tmp, reduction = "mds", group.by = "orig.ident") + 
                              scale_color_viridis(discrete = TRUE) + 
                              theme_void() + 
                              guides(fill = "none") + 
                              theme(plot.title = element_blank())
         
         MDS.plot.Heavy <- DimPlot(tmp, reduction = "mds", group.by = "Heavy.C") + 
                              scale_color_viridis(discrete = TRUE) + 
                              theme_void() + 
                              guides(fill = "none") + 
                              theme(plot.title = element_blank())
         
         tmp$Heavy.length <- nchar(stringr::str_split(tmp$CTaa, "_", simplify = TRUE)[,1])
         
         MDS.plot.HeavyL <- FeaturePlot(tmp, reduction = "mds", features = "Heavy.length") + 
                              scale_color_viridis() + 
                              theme_void() + 
                              guides(fill = "none") + 
                              theme(plot.title = element_blank())
         
         Dim.plot.HeavyL <- FeaturePlot(tmp, reduction = heavy[j], features = "Heavy.length") + 
                              scale_color_viridis() + 
                              theme_void() + 
                              guides(fill = "none") + 
                              theme(plot.title = element_blank())
         
          DimPlot(tmp, 
                         reduction = "mds", 
                         group.by = "Heavy.group", ) + 
                      scale_color_viridis(discrete = TRUE, option = "H", na.value = "grey") + 
                      theme_void() + 
                      theme(plot.title = element_blank()) 
          ggsave(paste0("./output/", heavy[j], "_cluster.pdf"), height = 3.5, width = 6)
          
          
         
         Dim.plot.H + Dim.plot.Ident + MDS.plot.H + MDS.plot.Ident + plot_layout(ncol = 2)
    ggsave(paste0("./output/", heavy[j], "_", paste0(c("P1.1", "P2.1", "HD1", "HD4"), collapse = "."), ".png"), height = 8, width = 8, dpi = 600)
    
     Dim.plot.HeavyL + MDS.plot.HeavyL + plot_layout(ncol = 2)
    ggsave(paste0("./output/", heavy[j], "_", paste0(c("P1.1", "P2.1", "HD1", "HD4"), collapse = "."), "_length.png"), height = 4, width = 8)
}
    
for (k in seq_along(light)) {
        Dim.plot.H <- DimPlot(seurat.merge[,cells], reduction = light[k], group.by = "Light.V") + 
                          scale_color_viridis(discrete = TRUE) + 
                          theme_void() + 
                          guides(fill = "none") + 
                          theme(plot.title = element_blank()) +
                          NoLegend()
        
        Dim.plot.Ident <- DimPlot(seurat.merge[,cells], reduction = light[k], group.by = "orig.ident") + 
                              scale_color_viridis(discrete = TRUE) + 
                              theme_void() + 
                              guides(fill = "none") + 
                              theme(plot.title = element_blank())
        tmp <- subset(seurat.merge, cells = rownames(seurat.merge[[]])[cells])
        
        #Calculate MDS  
        d <- dist(tmp@reductions[[light[k]]]@cell.embeddings)
        mds <- cmdscale(d = d, k = 2)
        colnames(mds) <- paste0("MDS_", 1:2)
        tmp[["mds"]] <- CreateDimReducObject(embeddings = mds, key = "MDS_", assay = DefaultAssay(tmp))
        
         
        MDS.plot.H <- DimPlot(tmp, reduction = "mds", group.by = "Light.V") + 
                          scale_color_viridis(discrete = TRUE) + 
                          theme_void() + 
                          guides(fill = "none") + 
                          theme(plot.title = element_blank()) +
                          NoLegend()
        
         MDS.plot.Ident <- DimPlot(tmp, reduction = "mds", group.by = "orig.ident") + 
                              scale_color_viridis(discrete = TRUE) + 
                              theme_void() + 
                              guides(fill = "none") + 
                              theme(plot.title = element_blank())
         
         MDS.plot.Light <- DimPlot(tmp, reduction = "mds", group.by = "Light.C") + 
                              scale_color_viridis(discrete = TRUE) + 
                              theme_void() + 
                              guides(fill = "none") + 
                              theme(plot.title = element_blank())
         
         tmp$Light.length <- nchar(stringr::str_split(tmp$CTaa, "_", simplify = TRUE)[,2])
         
         MDS.plot.LightL <- FeaturePlot(tmp, reduction = "mds", features = "Light.length") + 
                              scale_color_viridis() + 
                              theme_void() + 
                              guides(fill = "none") + 
                              theme(plot.title = element_blank())
         
         Dim.plot.LightL <- FeaturePlot(tmp, reduction = light[k], features = "Light.length") + 
                              scale_color_viridis() + 
                              theme_void() + 
                              guides(fill = "none") + 
                              theme(plot.title = element_blank())
         
         DimPlot(tmp, 
                         reduction = "mds", 
                         group.by = "Light.group", ) + 
                      scale_color_viridis(discrete = TRUE, option = "H", na.value = "grey") + 
                      theme_void() + 
                      theme(plot.title = element_blank()) + 
                              theme(plot.title = element_blank())
          ggsave(paste0("./output/", light[j], "_cluster.pdf"), height = 3.5, width = 8)

         
         Dim.plot.H + Dim.plot.Ident + MDS.plot.H + MDS.plot.Ident  + plot_layout(ncol = 2)
    ggsave(paste0("./output/", light[k], "_", paste0(c("P1.1", "P2.1", "HD1", "HD4"), collapse = "."), ".png"), height = 8, width = 8)
    
    Dim.plot.LightL + MDS.plot.LightL + plot_layout(ncol = 2)
    ggsave(paste0("./output/", light[k], "_", paste0(c("P1.1", "P2.1", "HD1", "HD4"), collapse = "."), "_length.png"), height = 4, width = 8)
}
```




```{r}
seurat.merge$Light.length <- nchar(stringr::str_split(seurat.merge$CTaa, "_", simplify = TRUE)[,2])
seurat.merge$Heavy.length <- nchar(stringr::str_split(seurat.merge$CTaa, "_", simplify = TRUE)[,1])


DimPlot(seurat.merge[,cells], reduction = "Ibex.L.KF", group.by = "Light.length") + 
                      scale_color_viridis(discrete = TRUE) + 
                      theme_void() + 
                      theme(plot.title = element_blank()) 
ggsave("./output/Ibex_light_cdr3length.pdf", height = 3.5, width = 5)

DimPlot(seurat.merge[,cells], reduction = "Ibex.H.KF", group.by = "Heavy.length") + 
                      scale_color_viridis(discrete = TRUE) + 
                      theme_void() + 
                      theme(plot.title = element_blank()) 
ggsave("./output/Ibex_heavy_cdr3length.pdf", height = 3.5, width = 5)
```


## Reducing the data set to clonotypes

We are selecting only the P1.1 and P2.1 samples as they have GEX, ADT, and TCR sequencing.

```{r}
cells <- which(seurat.merge[[]]$orig.ident %in% c("P1.1", "P2.1"))

subset.Obj <- subset(seurat.merge, cells = rownames(seurat.merge[[]][cells,])) 

subset.Obj <- RunPCA(subset.Obj, verbose = FALSE)

subset.Obj <- NormalizeData(subset.Obj, assay = "ADT", normalization.method = 'CLR', margin = 2) 
IG.ADT <- grep("Ig",  rownames(subset.Obj[["ADT"]]))
VariableFeatures(subset.Obj, assay = "ADT") <- rownames(subset.Obj[["ADT"]])[-IG.ADT] #Removing Immunoglobulins from DimRed
subset.Obj <- ScaleData(subset.Obj, assay = "ADT")
subset.Obj <- RunPCA(subset.Obj, reduction.name = 'apca', assay = "ADT")

conga <- CoNGAfy(subset.Obj, 
                 assay = c("RNA", "ADT"))

#Get count of number of clones
meta <- seurat.merge[[]]
clone.count <- meta %>%
  group_by(CTaa) %>%
  summarise(n = n()) %>%
  as.data.frame()

ID <- clone.count$CTaa
clone.count <- data.frame(n = clone.count[,2])
rownames(clone.count) <- ID
conga <- AddMetaData(conga, clone.count)

meta <- subset.Obj[[]]
CTaa.unique <- unique(meta$CTaa)
for(i in seq_along(CTaa.unique)) {
  meta.tmp <- meta[meta$CTaa == CTaa.unique[i],]
  CTaa.tmp <- c(CTaa = unique(meta.tmp$CTaa), 
                donor = (paste(str_sort(unique(meta.tmp$orig.ident)), collapse = ";")), 
                H.V = paste(str_sort(unique(meta.tmp$Heavy.V)), collapse = ";"), 
                H.C = paste(str_sort(unique(meta.tmp$Heavy.C)), collapse = ";"), 
                L.V = paste(str_sort(unique(meta.tmp$Light.V)), collapse = ";"), 
                L.C = paste(str_sort(unique(meta.tmp$Light.C)), collapse = ";"),
                Heavy.group = paste(str_sort(unique(meta.tmp$Heavy.group)), collapse = ";"), 
                Light.group = paste(str_sort(unique(meta.tmp$Light.group)), collapse = ";"), 
                status <- paste(str_sort(unique(meta.tmp$status)), collapse = ";"))
  if(i == 1) {
    CTaa.Library <- CTaa.tmp
  } else {
    CTaa.Library <- rbind(CTaa.Library, CTaa.tmp)
  }
}
CTaa.Library <- as.data.frame(CTaa.Library)
aa <- CTaa.Library$CTaa
CTaa.Library <- CTaa.Library[,-1]
rownames(CTaa.Library) <- aa
colnames(CTaa.Library)[8] <- c("status")

conga <- AddMetaData(conga, CTaa.Library)

DefaultAssay(conga) <- 'RNA'
conga <- NormalizeData(conga) %>% 
  FindVariableFeatures() %>% 
  quietBCRgenes() %>% 
  ScaleData() %>% 
  RunPCA(verbose = FALSE)

#Processing ADT
DefaultAssay(conga) <- 'ADT'
IG.ADT <- grep("Ig",  rownames(conga[["ADT"]]))
VariableFeatures(conga) <- rownames(conga[["ADT"]])[-IG.ADT] #Removing Immunoglobulins from DimRed

conga <- NormalizeData(conga, normalization.method = 'CLR', margin = 2) %>% 
  ScaleData() %>% 
  RunPCA(reduction.name = 'apca')

##################
#Running Ibex
################
conga <- runIbex(conga, 
                    chains = "Heavy",
                    AA.properties = "KF", 
                    reduction.name = "Ibex.H.KF")

conga <- runIbex(conga, 
                    chains = "Light",
                    AA.properties = "KF", 
                    reduction.name = "Ibex.L.KF")



###################
#Trimodal Embedding
###################
output <- rescaleByNeighbors(
                  list(conga[["Ibex.H.KF"]]@cell.embeddings, 
                      conga[["pca"]]@cell.embeddings, 
                      conga[["apca"]]@cell.embeddings), 
                  k = 50)

conga[["Trimodal.Embed"]] <- CreateDimReducObject(
                                embeddings = as.matrix(output),
                                stdev = rep(0, ncol(output)),
                                key = "Combined",
                                jackstraw = NULL,
                                misc = list())

conga <- FindNeighbors(conga, 
                       reduction = "Trimodal.Embed", 
                       dims = 1:130, 
                       graph.name = c("Ibex_nn", "Ibex_snn"))
conga <- FindClusters(conga,
                      graph.name = "Ibex_nn",
                      algorithm = 4)

#######################################
#UMAP Comparisons between each approach
#######################################

conga <- RunUMAP(conga, 
                     reduction = 'pca', 
                     dims = 1:30, 
                     assay = 'RNA', 
                     reduction.name = 'rna.umap', 
                     reduction.key = 'rnaUMAP_')

conga <- RunUMAP(conga, 
                     reduction = 'apca', 
                     dims = 1:30, 
                     assay = 'ADT', 
                     reduction.name = 'adt.umap', 
                     reduction.key = 'adtUMAP_')

conga <- RunUMAP(conga, 
                     reduction = 'Ibex.H.KF', 
                     dims = 1:30, 
                     reduction.name = 'heavy.umap', 
                     reduction.key = 'heavyUMAP_')

conga <- RunUMAP(conga, 
                     reduction = 'Ibex.L.KF', 
                     dims = 1:30, 
                     reduction.name = 'light.umap', 
                     reduction.key = 'lightUMAP_')

conga <- RunUMAP(conga, 
                     reduction = 'Trimodal.Embed', 
                     dims = 1:130, 
                     reduction.name = 'trimodal.umap', 
                     reduction.key = 'trimodalUMAP_')

plot.rna <- DimPlot(conga, reduction = "rna.umap") + NoLegend()
plot.adt <- DimPlot(conga, reduction = "adt.umap") + NoLegend()
plot.heavy <- DimPlot(conga, reduction = "heavy.umap") + NoLegend()
plot.combined <- DimPlot(conga, reduction = "trimodal.umap") + NoLegend()

###############
#Plotting
#############

plot.rna + 
  scale_color_tableau() + 
  theme_void() + 
  NoLegend()
ggsave("./output/clonotype_RNA_umap.pdf", height = 3.5, width = 3.5)

plot.adt + 
  scale_color_tableau() + 
  theme_void() + 
  NoLegend()
ggsave("./output/clonotype_ADT_umap.pdf", height = 3.5, width = 3.5)

plot.heavy + 
  scale_color_tableau() + 
  theme_void() + 
  NoLegend()
ggsave("./output/clonotype_IgHeavy_umap.pdf", height = 3.5, width = 3.5)

plot.combined + 
  scale_color_tableau() + 
  theme_void() + 
  NoLegend()
ggsave("./output/clonotype_combined_umap.pdf", height = 3.5, width = 3.5)

CD45.plot <- FeaturePlot(conga, 
                 reduction = "trimodal.umap", 
                 features = c("CD45RA", "CD45RO"), 
                 cols = viridis_pal()(50), 
                 combine = FALSE) 

for(i in seq_along(CD45.plot)) {
  CD45.plot[[i]] <- CD45.plot[[i]] + 
              guides(color = "none") + 
              theme_void() + 
              theme(plot.title = element_blank(), 
                    plot.margin = unit(c(0.6, 0.15, 0.6, 0.15), "cm"))
}
plot_a_list(CD45.plot,2,1)
ggsave("./output/combined_CD45plots.png", height = 10, width = 5, dpi = 600)

saveRDS(conga, file = "./data/Ibex_ClonotypeRep.rds")
```

# Conclusions

```{r}
sessionInfo()
```


